### ISSUE TITLE
StableDiffusionControlNetPipeline: VRAM exploding up to 1 GB every single step

### ISSUE THREAD

**Issue by oxysoft** [huggingface/diffusers#9600]
### Describe the bug

It seems that when I run ControlNet, the VRAM explodes out of control. I'm not sure if it happens with just one ControlNet. No problem if I run the basic StableDiffusionPipeline without any ControlNet.

### Reproduction

```
# !pip install opencv-python transformers accelerate
from diffusers import StableDiffusionControlNetPipeline, ControlNetModel, UniPCMultistepScheduler
from diffusers.utils import load_image
import numpy as np
import torch

import cv2
from PIL import Image

# download an image
image = load_image(
    "https://hf.co/datasets/huggingface/documentation-images/resolve/main/diffusers/input_image_vermeer.png"
)
image = np.array(image)

# get canny image
image = cv2.Canny(image, 100, 200)
image = image[:, :, None]
image = np.concatenate([image, image, image], axis=2)
canny_image = Image.fromarray(image)

# load control net and stable diffusion v1-5
cn1 = ControlNetModel.from_pretrained("lllyasviel/control_v11p_sd15_softedge", torch_dtype=torch.float16)
cn2 = ControlNetModel.from_pretrained("CiaraRowles/TemporalNet", torch_dtype=torch.float16)
cn3 = ControlNetModel.from_pretrained("lllyasviel/sd-controlnet-depth", torch_dtype=torch.float16)

pipe = StableDiffusionControlNetPipeline.from_pretrained(
	"SG161222/Realistic_Vision_V1.4",
	controlnet=[cn1, cn2, cn3],
	torch_dtype=torch.float16
)

# speed up diffusion process with faster scheduler and memory optimization
pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config)

pipe.enable_model_cpu_offload()

# generate image
generator = torch.manual_seed(0)

def on_step_end(pipeline, i, t, callback_kwargs):
	import subprocess

	def get_gpu_memory_usage():
		result = subprocess.run(['nvidia-smi', '--query-gpu=memory.used', '--format=csv,nounits,noheader'], stdout=subprocess.PIPE, check=True)
		return int(result.stdout.decode('utf-8').strip())

	vram_usage = get_gpu_memory_usage()
	print(f"Current VRAM usage: {vram_usage} MB")
	return callback_kwargs

for i in range(10):
	print(f"Iteration {i}")
	image = pipe(
		"futuristic-looking woman",
		num_inference_steps=20,
		generator=generator,
		image=[canny_image, canny_image, canny_image],
		controlnet_guidance_scale=[0.2, 0.2, 0.2],
		callback_on_step_end=on_step_end
	).images[0]
```

### Logs

```shell
/media/data-team/Projects/discore/venv/bin/python -X pycache_prefix=/home/nuck/.cache/JetBrains/PyCharmCE2023.3/cpython-cache /home/nuck/.local/share/JetBrains/Toolbox/apps/pycharm-community/plugins/python-ce/helpers/pydev/pydevd.py --multiprocess --qt-support=auto --client 127.0.0.1 --port 37689 --file /media/data-team/Projects/discore/repro.py 
Connected to pydev debugger (build 233.13763.11)
/home/nuck/.local/share/JetBrains/Toolbox/apps/pycharm-community/plugins/python-ce/helpers/pydev/_pydevd_bundle/pydevd_pep_669_tracing.py:41: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
  while frame and isinstance(frame.f_locals.get('self'), PEP669CallbackBase):
unet/diffusion_pytorch_model.safetensors not found
Loading pipeline components...:   0%|          | 0/7 [00:00<?, ?it/s]An error occurred while trying to fetch /media/data-team/Caches/huggingface/hub/models--SG161222--Realistic_Vision_V1.4/snapshots/686d3dcb8bbc7e6a7757fd161e0fbafd23d6c629/vae: Error no file named diffusion_pytorch_model.safetensors found in directory /media/data-team/Caches/huggingface/hub/models--SG161222--Realistic_Vision_V1.4/snapshots/686d3dcb8bbc7e6a7757fd161e0fbafd23d6c629/vae.
Defaulting to unsafe serialization. Pass `allow_pickle=False` to raise an error instead.
Loading pipeline components...:  14%|█▍        | 1/7 [00:01<00:07,  1.27s/it]/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/transformers/tokenization_utils_base.py:1617: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be deprecated in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(
Loading pipeline components...:  57%|█████▋    | 4/7 [00:03<00:02,  1.33it/s]An error occurred while trying to fetch /media/data-team/Caches/huggingface/hub/models--SG161222--Realistic_Vision_V1.4/snapshots/686d3dcb8bbc7e6a7757fd161e0fbafd23d6c629/unet: Error no file named diffusion_pytorch_model.safetensors found in directory /media/data-team/Caches/huggingface/hub/models--SG161222--Realistic_Vision_V1.4/snapshots/686d3dcb8bbc7e6a7757fd161e0fbafd23d6c629/unet.
Defaulting to unsafe serialization. Pass `allow_pickle=False` to raise an error instead.
Loading pipeline components...: 100%|██████████| 7/7 [00:18<00:00,  2.64s/it]
Iteration 0
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
  5%|▌         | 1/20 [00:01<00:32,  1.71s/it]Current VRAM usage: 6439 MB
Current VRAM usage: 6855 MB
 15%|█▌        | 3/20 [00:02<00:12,  1.36it/s]Current VRAM usage: 7349 MB
Current VRAM usage: 7739 MB
 25%|██▌       | 5/20 [00:03<00:08,  1.82it/s]Current VRAM usage: 8163 MB
Current VRAM usage: 8663 MB
 35%|███▌      | 7/20 [00:04<00:06,  2.07it/s]Current VRAM usage: 9161 MB
Current VRAM usage: 9613 MB
 45%|████▌     | 9/20 [00:05<00:04,  2.21it/s]Current VRAM usage: 10153 MB
Current VRAM usage: 10653 MB
 55%|█████▌    | 11/20 [00:06<00:03,  2.27it/s]Current VRAM usage: 11055 MB
Current VRAM usage: 11455 MB
 65%|██████▌   | 13/20 [00:06<00:03,  2.28it/s]Current VRAM usage: 11893 MB
Current VRAM usage: 12293 MB
 75%|███████▌  | 15/20 [00:07<00:02,  2.30it/s]Current VRAM usage: 12761 MB
Current VRAM usage: 13189 MB
 85%|████████▌ | 17/20 [00:08<00:01,  2.28it/s]Current VRAM usage: 13587 MB
Current VRAM usage: 13997 MB
 95%|█████████▌| 19/20 [00:09<00:00,  2.30it/s]Current VRAM usage: 14471 MB
Current VRAM usage: 14923 MB
100%|██████████| 20/20 [00:09<00:00,  2.01it/s]
Iteration 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
  5%|▌         | 1/20 [00:01<00:26,  1.37s/it]Current VRAM usage: 6479 MB
Current VRAM usage: 6877 MB
 15%|█▌        | 3/20 [00:02<00:10,  1.58it/s]Current VRAM usage: 7253 MB
Current VRAM usage: 7703 MB
 25%|██▌       | 5/20 [00:03<00:07,  1.97it/s]Current VRAM usage: 8227 MB
Current VRAM usage: 8609 MB
 35%|███▌      | 7/20 [00:03<00:05,  2.18it/s]Current VRAM usage: 9071 MB
Current VRAM usage: 9519 MB
 45%|████▌     | 9/20 [00:04<00:04,  2.26it/s]Current VRAM usage: 9983 MB
Current VRAM usage: 10449 MB
 55%|█████▌    | 11/20 [00:05<00:03,  2.30it/s]Current VRAM usage: 10913 MB
Current VRAM usage: 11327 MB
 65%|██████▌   | 13/20 [00:06<00:03,  2.28it/s]Current VRAM usage: 11789 MB
Current VRAM usage: 12243 MB
 75%|███████▌  | 15/20 [00:07<00:02,  2.30it/s]Current VRAM usage: 12651 MB
Current VRAM usage: 13215 MB
 85%|████████▌ | 17/20 [00:08<00:01,  2.32it/s]Current VRAM usage: 13743 MB
Current VRAM usage: 14225 MB
 95%|█████████▌| 19/20 [00:09<00:00,  2.34it/s]Current VRAM usage: 14667 MB
Current VRAM usage: 15121 MB
100%|██████████| 20/20 [00:09<00:00,  2.11it/s]
Iteration 2
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
  5%|▌         | 1/20 [00:01<00:26,  1.40s/it]Current VRAM usage: 15841 MB
Current VRAM usage: 16307 MB
 15%|█▌        | 3/20 [00:02<00:11,  1.53it/s]Current VRAM usage: 16719 MB
Current VRAM usage: 17235 MB
 25%|██▌       | 5/20 [00:03<00:07,  1.92it/s]Current VRAM usage: 17657 MB
Current VRAM usage: 18141 MB
 35%|███▌      | 7/20 [00:04<00:06,  2.12it/s]Current VRAM usage: 18607 MB
Current VRAM usage: 19007 MB
 45%|████▌     | 9/20 [00:04<00:04,  2.25it/s]Current VRAM usage: 19419 MB
Current VRAM usage: 19777 MB
 55%|█████▌    | 11/20 [00:06<00:05,  1.70it/s]Current VRAM usage: 6867 MB
Current VRAM usage: 6789 MB
 65%|██████▌   | 13/20 [00:07<00:03,  1.96it/s]Current VRAM usage: 7313 MB
Current VRAM usage: 7723 MB
 75%|███████▌  | 15/20 [00:07<00:02,  2.13it/s]Current VRAM usage: 8127 MB
Current VRAM usage: 8517 MB
 85%|████████▌ | 17/20 [00:08<00:01,  2.23it/s]Current VRAM usage: 8911 MB
Current VRAM usage: 9441 MB
 95%|█████████▌| 19/20 [00:09<00:00,  2.28it/s]Current VRAM usage: 9869 MB
Current VRAM usage: 10363 MB
100%|██████████| 20/20 [00:10<00:00,  1.98it/s]
Iteration 3
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
  5%|▌         | 1/20 [00:01<00:26,  1.38s/it]Current VRAM usage: 10931 MB
Current VRAM usage: 11305 MB
 15%|█▌        | 3/20 [00:02<00:10,  1.57it/s]Current VRAM usage: 11733 MB
Current VRAM usage: 12103 MB
 25%|██▌       | 5/20 [00:03<00:07,  1.98it/s]Current VRAM usage: 12455 MB
Current VRAM usage: 12983 MB
 35%|███▌      | 7/20 [00:03<00:06,  2.10it/s]Current VRAM usage: 13405 MB
Current VRAM usage: 13975 MB
 45%|████▌     | 9/20 [00:04<00:05,  2.17it/s]Current VRAM usage: 14445 MB
Current VRAM usage: 14869 MB
 55%|█████▌    | 11/20 [00:05<00:04,  2.25it/s]Current VRAM usage: 15299 MB
Current VRAM usage: 15689 MB
 65%|██████▌   | 13/20 [00:06<00:03,  2.31it/s]Current VRAM usage: 16145 MB
Current VRAM usage: 16661 MB
 75%|███████▌  | 15/20 [00:07<00:02,  2.32it/s]Current VRAM usage: 17179 MB
Current VRAM usage: 17667 MB
 85%|████████▌ | 17/20 [00:08<00:01,  2.33it/s]Current VRAM usage: 18155 MB
Current VRAM usage: 18601 MB
 95%|█████████▌| 19/20 [00:09<00:00,  2.31it/s]Current VRAM usage: 19073 MB
Current VRAM usage: 19567 MB
100%|██████████| 20/20 [00:09<00:00,  2.08it/s]
Iteration 4
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
  5%|▌         | 1/20 [00:01<00:32,  1.72s/it]Current VRAM usage: 6275 MB
Current VRAM usage: 6775 MB
 15%|█▌        | 3/20 [00:02<00:12,  1.39it/s]Current VRAM usage: 7265 MB
Current VRAM usage: 7713 MB
 25%|██▌       | 5/20 [00:03<00:08,  1.82it/s]Current VRAM usage: 8163 MB
Current VRAM usage: 8581 MB
 35%|███▌      | 7/20 [00:04<00:06,  2.06it/s]Current VRAM usage: 9057 MB
Current VRAM usage: 9667 MB
 45%|████▌     | 9/20 [00:05<00:04,  2.21it/s]Current VRAM usage: 10113 MB
Current VRAM usage: 10595 MB
 55%|█████▌    | 11/20 [00:06<00:03,  2.27it/s]Current VRAM usage: 10977 MB
Current VRAM usage: 11503 MB
 65%|██████▌   | 13/20 [00:06<00:03,  2.28it/s]Current VRAM usage: 11963 MB
Current VRAM usage: 12319 MB
 75%|███████▌  | 15/20 [00:07<00:02,  2.31it/s]Current VRAM usage: 12791 MB
Current VRAM usage: 13211 MB
 85%|████████▌ | 17/20 [00:08<00:01,  2.33it/s]Current VRAM usage: 13637 MB
Current VRAM usage: 14077 MB
 95%|█████████▌| 19/20 [00:09<00:00,  2.34it/s]Current VRAM usage: 14467 MB
Current VRAM usage: 14897 MB
100%|██████████| 20/20 [00:09<00:00,  2.02it/s]
Iteration 5
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
  5%|▌         | 1/20 [00:01<00:26,  1.41s/it]Current VRAM usage: 15417 MB
Current VRAM usage: 15765 MB
 15%|█▌        | 3/20 [00:02<00:11,  1.53it/s]Current VRAM usage: 16239 MB
Current VRAM usage: 16763 MB
 25%|██▌       | 5/20 [00:03<00:07,  1.93it/s]Current VRAM usage: 17173 MB
Current VRAM usage: 17659 MB
 35%|███▌      | 7/20 [00:04<00:06,  2.14it/s]Current VRAM usage: 18013 MB
Current VRAM usage: 18473 MB
 45%|████▌     | 9/20 [00:04<00:04,  2.25it/s]Current VRAM usage: 19041 MB
Current VRAM usage: 19451 MB
 55%|█████▌    | 11/20 [00:05<00:04,  2.04it/s]Current VRAM usage: 19623 MB
Current VRAM usage: 6675 MB
 65%|██████▌   | 13/20 [00:07<00:03,  1.90it/s]Current VRAM usage: 6963 MB
Current VRAM usage: 7317 MB
 75%|███████▌  | 15/20 [00:07<00:02,  2.11it/s]Current VRAM usage: 7665 MB
Current VRAM usage: 8073 MB
 85%|████████▌ | 17/20 [00:08<00:01,  2.19it/s]Current VRAM usage: 8537 MB
Current VRAM usage: 8883 MB
 95%|█████████▌| 19/20 [00:09<00:00,  2.23it/s]Current VRAM usage: 9385 MB
Current VRAM usage: 9801 MB
100%|██████████| 20/20 [00:10<00:00,  1.98it/s]
Iteration 6
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
  5%|▌         | 1/20 [00:01<00:26,  1.37s/it]Current VRAM usage: 10263 MB
Current VRAM usage: 10635 MB
 15%|█▌        | 3/20 [00:02<00:10,  1.55it/s]Current VRAM usage: 11165 MB
Current VRAM usage: 11579 MB
 25%|██▌       | 5/20 [00:03<00:07,  1.94it/s]Current VRAM usage: 12075 MB
Current VRAM usage: 12471 MB
 35%|███▌      | 7/20 [00:03<00:06,  2.14it/s]Current VRAM usage: 12935 MB
Current VRAM usage: 13329 MB
 45%|████▌     | 9/20 [00:04<00:04,  2.26it/s]Current VRAM usage: 13655 MB
Current VRAM usage: 14145 MB
 55%|█████▌    | 11/20 [00:05<00:03,  2.32it/s]Current VRAM usage: 14635 MB
Current VRAM usage: 15069 MB
 65%|██████▌   | 13/20 [00:06<00:02,  2.34it/s]Current VRAM usage: 15507 MB
Current VRAM usage: 15947 MB
 75%|███████▌  | 15/20 [00:07<00:02,  2.34it/s]Current VRAM usage: 16295 MB
Current VRAM usage: 16736 MB
 85%|████████▌ | 17/20 [00:08<00:01,  2.31it/s]Current VRAM usage: 17116 MB
Current VRAM usage: 17588 MB
 95%|█████████▌| 19/20 [00:09<00:00,  2.29it/s]Current VRAM usage: 18062 MB
Current VRAM usage: 18274 MB
100%|██████████| 20/20 [00:09<00:00,  2.10it/s]
Iteration 7
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
False, <PIL.Image.Image image mode=RGB size=512x512 at 0x7F201BBBAE70>, <class 'PIL.Image.Image'>, 1
Traceback (most recent call last):
  File "/media/data-team/Projects/discore/repro.py", line 54, in <module>
    image = pipe(
            ^^^^^
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/diffusers/pipelines/controlnet/pipeline_controlnet.py", line 1119, in __call__
    prompt_embeds, negative_prompt_embeds = self.encode_prompt(
                                            ^^^^^^^^^^^^^^^^^^^
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/diffusers/pipelines/controlnet/pipeline_controlnet.py", line 382, in encode_prompt
    prompt_embeds = self.text_encoder(text_input_ids.to(device), attention_mask=attention_mask)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/accelerate/hooks.py", line 165, in new_forward
    args, kwargs = module._hf_hook.pre_forward(module, *args, **kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/accelerate/hooks.py", line 700, in pre_forward
    module.to(self.execution_device)
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/transformers/modeling_utils.py", line 2958, in to
    return super().to(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1174, in to
    return self._apply(convert)
           ^^^^^^^^^^^^^^^^^^^^
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 780, in _apply
    module._apply(fn)
  [Previous line repeated 3 more times]
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 804, in _apply
    with torch.no_grad():
  File "/media/data-team/Projects/discore/venv/lib/python3.12/site-packages/torch/autograd/grad_mode.py", line 84, in __exit__
    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
    
KeyboardInterrupt
```

### System Info

```
- 🤗 Diffusers version: 0.30.3
- Platform: Linux-6.6.52-1-MANJARO-x86_64-with-glibc2.40
- Running on Google Colab?: No
- Python version: 3.12.5
- PyTorch version (GPU?): 2.4.1+cu121 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Huggingface_hub version: 0.25.1
- Transformers version: 4.45.1
- Accelerate version: 0.34.2
- PEFT version: not installed
- Bitsandbytes version: not installed
- Safetensors version: 0.4.5
- xFormers version: not installed
- Accelerator: NVIDIA GeForce RTX 3090, 24576 MiB
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: single gpu
```

### Who can help?

@sayakpaul @yiyixuxu

**Comment by dummyuser-123** [comment]
You have to release the memory after each iterations, so that it will not store the extra memory. you can add this code after completion of one diffusion process.

def flush_memory():
	gc.collect()
	if torch.cuda.is_available():
		torch.cuda.empty_cache()
		torch.cuda.ipc_collect()

**Comment by oxysoft** [comment]
That seems to work, but I'm pretty sure it didn't use to do this. I'm running my same usual codebase for AI animation and it's now doing this whereas this codebase had no problem in the past. (which was many months ago on completely different versions of torch, other pip packages, python, etc.)

Climbing by over 10gb over the course of 20 steps makes no sense to me, I would expect the tensors to be deleted or reused so it's a flat increase no matter how many steps are run. With other models loaded, I will have to clear the RAM not between iteration but between every step, because I may only have a couple gigabytes left. Running multiple controlnet has never required a whole 16GB+ of VRAM like this, I used to run this on a 6GB card.

**Comment by oxysoft** [comment]
Alright as I predicted there was something wrong with my packages. I really couldn't say what exactly, but maybe this can set somebody on the right track in the future if this comes up again.

I think I can at least trace it down to this sequence of operations: the demucs package (model for music source separation) is stuck on some older versions of torch or numpy, but I figured out that the difference is trivial, and the newest packages can be used with no problem to demucs. So I had installed demucs first so it pulls in all its requirements, then installed my own packages on top I believe, upgrading versions in some cases.

To fix it just now, I simply uninstalled torch, torchvision, and torchaudio, and reinstalled them. I also had torch+cu121 while I'm my system is currently on CUDA 12.4 (as reported by nvidia-smi).

I don't know how any of this correlates to the VRAM exploding, I would've expected more serious bugs or crashes if there were incompatibilities, but in any case it works as expected now.

No regression in Diffusers!
